### PR TITLE
treewide: change assert() to SCYLLA_ASSERT()

### PR DESCRIPTION
**Pull request by avikivity**
assert() is traditionally disabled in release builds, but not in scylladb. This hasn't caused problems so far, but the latest abseil release includes a commit [1] that causes a 1000 insn/op regression when NDEBUG is not defined.

Clearly, we must move towards a build system where NDEBUG is defined in release builds. But we can't just define it blindly without vetting all the assert() calls, as some were written with the expectation that they are enabled in release mode.

To solve the conundrum, change all assert() calls to a new SCYLLA_ASSERT() macro in utils/assert.hh. This macro is always defined and is not conditional on NDEBUG, so we can later (after vetting Seastar) enable NDEBUG in release mode.

[1] https://github.com/abseil/abseil-cpp/commit/66ef711d6846771b8e725e377de37bedae6c1527

No backport needed, this is to prepare ScyllaDB for the wider ecosystem.

### CODE REVIEW

**Comment by denesb** [comment]
Is the change entirely mechanical? Or are there "organic" changes?

**Comment by avikivity** [comment]
> Is the change entirely mechanical? Or are there "organic" changes?

utils/assert.hh was written by a human

**Review by denesb — APPROVED** [review]
I only reviewed the part that was written by a human -- `utils/assert.hh`.
I'll leave the review of the mechanical changes up to ~~AI~~ the compiler.

**Review by nyh — APPROVED** [review]
Looks good, although when the dust settles, I don't think anyone will remember why we have SCYLLA_ASSERT() and how it differs from the regular assert(). Perhaps it should have a more self-explanatory name like always_assert(), or if we keep the name SCYLLA_ASSERT() (I'm not against it), please add a comment in assert.hh explaining how it differs from the regular assert().

**Comment by avikivity** [comment]
> Looks good, although when the dust settles, I don't think anyone will remember why we have SCYLLA_ASSERT() and how it differs from the regular assert(). Perhaps it should have a more self-explanatory name like always_assert(), or if we keep the name SCYLLA_ASSERT() (I'm not against it), please add a comment in assert.hh explaining how it differs from the regular assert().

I'll add a comment. always_assert() isn't a good name, because we may want to change the behavior.

**Comment by avikivity** [comment]
v2: documented SCYLLA_ASSERT()

File: utils/assert.hh [hunk 1]
```diff
@@ -0,0 +1,9 @@
+// Copyright 2024-present ScyllaDB
+// SPDX-License-Identifier: AGPL-3.0-or-later
+
+#pragma once
+
+#include <cassert>
+
+/// Like assert(), but indpendent of NDEBUG. Active in all release modes.
```
Comment (nyh) [inline utils/assert.hh:8]: Typo "indpendent" and I think you meant all "build modes", not "release modes"?

**Review by nyh — CHANGES_REQUESTED** [review]

**Comment by avikivity** [comment]
v3:
 - fix type in "independent"
 - build modes, not release modes

**Pull request closed without merging**